src/app/clusters/time-synchronization-server/time-synchronization-server.cpp

```diff
@@ -0,0 +1,685 @@
+                                                                         : mDefaultNtp.SetNull();
+
+    auto tz = mTimeZoneList.begin();
+    for (size_t i = 0; i < mTimeZoneList.size(); i++)
+    {
+        const char * buf = reinterpret_cast<const char *>(mNames[i].name);
+        tz[i].name.SetValue(chip::CharSpan(buf, sizeof(mNames[i].name)));
+    }
+    mTimeSyncDataProvider.LoadTimeZone(mTimeZoneList);
+    mTimeSyncDataProvider.LoadDSTOffset(mDstOffList);
+    // TODO if trusted time source is available schedule a time read
+    if (!mTrustedTimeSource.IsNull())
+    {
+    }
+}
+
+void TimeSynchronizationServer::ScheduleDelayedAction(System::Clock::Seconds32 delay, System::TimerCompleteCallback action,
+                                                      void * aAppState)
+{
+    VerifyOrDie(SystemLayer().StartTimer(std::chrono::duration_cast<System::Clock::Timeout>(delay), action, aAppState) ==
+                CHIP_NO_ERROR);
+}
+
+namespace {
+
+struct TimeZoneCodec
+{
+    static constexpr TLV::Tag TagOffset() { return TLV::ContextTag(TimeSynchronization::Structs::TimeZoneStruct::Fields::kOffset); }
+    static constexpr TLV::Tag TagValidAt()
+    {
+        return TLV::ContextTag(TimeSynchronization::Structs::TimeZoneStruct::Fields::kValidAt);
+    }
+    static constexpr TLV::Tag TagName() { return TLV::ContextTag(TimeSynchronization::Structs::TimeZoneStruct::Fields::kName); }
+
+    TimeSynchronization::Structs::TimeZoneStruct::Type timeZone;
+
+    TimeZoneCodec(TimeSynchronization::Structs::TimeZoneStruct::Type tz) : timeZone(tz) {}
+
+    static constexpr bool kIsFabricScoped = false;
+
+    CHIP_ERROR Encode(TLV::TLVWriter & writer, TLV::Tag tag) const
+    {
+        TLV::TLVType outer;
+        ReturnErrorOnFailure(writer.StartContainer(tag, TLV::kTLVType_Structure, outer));
+
+        // Offset
+        ReturnErrorOnFailure(DataModel::Encode(writer, TagOffset(), timeZone.offset));
+        // ValidAt
+        ReturnErrorOnFailure(DataModel::Encode(writer, TagValidAt(), timeZone.validAt));
+        // Name
+        if (timeZone.name.HasValue())
+        {
+            uint32_t name_size = static_cast<uint32_t>(strnlen(timeZone.name.Value().data(), 64));
+            ReturnErrorOnFailure(writer.PutString(TagName(), timeZone.name.Value().data(), name_size));
+            ChipLogProgress(Zcl, "%s %d", timeZone.name.Value().data(), name_size);
+        }
+        ReturnErrorOnFailure(writer.EndContainer(outer));
+        return CHIP_NO_ERROR;
+    }
+};
+
+static bool computeLocalTime(chip::EndpointId ep)
+{
+    DataModel::Nullable<uint64_t> utcTime, localTime;
+    int32_t timeZoneOffset = 0, dstOffset = 0;
+    UTCTime::Get(ep, utcTime);
+    if (utcTime.IsNull())
+    {
+        return false;
+    }
+    auto tz  = TimeSynchronizationServer::Instance().GetTimeZone().begin();
+    auto dst = TimeSynchronizationServer::Instance().GetDSTOffset().begin();
+    if (tz->validAt <= utcTime.Value())
```

Comment:
Does this code purge the expired values from the tz and dst tables? If not, or if that's not guaranteed to run before this code, then you have to check that the next value in the list isn't the one that's valid. Ditto for DST.

Reply:
The assumption here is that expired tz & dst get purged. That whole logic for both tz and dst is not available yet but this code assumes that. Need to be taken care of.